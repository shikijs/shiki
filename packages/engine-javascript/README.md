# @shikijs/engine-javascript

Engine for Shiki using JavaScript's native RegExp (experimental).

[Documentation](https://shiki.style/guide/regex-engines)

## License

MIT
