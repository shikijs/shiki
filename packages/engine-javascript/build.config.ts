import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
  entries: [
    'src/index.ts',
  ],
  declaration: true,
  rollup: {
    emitCJS: false,
    dts: {
      compilerOptions: {
        paths: {},
      },
    },
  },
  externals: [
    'hast',
  ],
})
