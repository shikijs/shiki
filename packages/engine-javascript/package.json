{
  "name": "@shikijs/engine-javascript",
  "type": "module",
  "version": "1.16.3",
  "description": "Engine for Shiki using JavaScript's native RegExp",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/shikijs/shiki#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/shikijs/shiki.git",
    "directory": "packages/transformers"
  },
  "bugs": "https://github.com/shikijs/shiki/issues",
  "keywords": [
    "shiki",
    "@shikijs/transformers"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.mts",
      "default": "./dist/index.mjs"
    }
  },
  "main": "./dist/index.mjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.mts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "unbuild",
    "dev": "unbuild --stub",
    "prepublishOnly": "nr build"
  },
  "dependencies": {
    "@shikijs/types": "workspace:*",
    "oniguruma-to-js": "catalog:",
    "regex": "catalog:"
  }
}
