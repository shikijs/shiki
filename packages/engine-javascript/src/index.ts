export * from './engine-compile'
export * from './engine-raw'
export * from './scanner'
