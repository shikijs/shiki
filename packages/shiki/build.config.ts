import fs from 'node:fs/promises'
import fg from 'fast-glob'
import { basename, dirname, join } from 'pathe'
import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
  entries: [
    'src/index.ts',
    'src/core.ts',
    'src/core-unwasm.ts',
    'src/types.ts',
    'src/themes.ts',
    'src/langs.ts',
    'src/wasm.ts',
    'src/bundle-full.ts',
    'src/bundle-web.ts',
    'src/engine-javascript.ts',
    'src/engine-oniguruma.ts',
    'src/textmate.ts',
  ],
  externals: [
    'shiki/wasm',
    '@shikijs/types',
    'hast',
    /^@shikijs[/\\].*/g,
    /[/\\](langs|themes)[/\\]/g,
  ],
  declaration: 'node16',
  clean: true,
  hooks: {
    'rollup:build': async () => {
      await fs.cp('./src/langs', './dist/langs', { recursive: true })
      await fs.cp('./src/themes', './dist/themes', { recursive: true })

      await fs.copyFile(
        './node_modules/vscode-oniguruma/release/onig.wasm',
        './dist/onig.wasm',
      )

      await fs.writeFile('dist/onig.d.mts', 'declare const binary: ArrayBuffer; export default binary;', 'utf-8')
      const langs = await fg('dist/langs/*.mjs', { absolute: true })
      await Promise.all(
        langs.map(file => fs.writeFile(
          join(dirname(file), `${basename(file, '.mjs')}.d.mts`),
          'import { LanguageRegistration } from \'@shikijs/core\';declare const reg: LanguageRegistration[];export default reg',
          'utf-8',
        )),
      )
      const themes = await fg('dist/themes/*.mjs', { absolute: true })
      await Promise.all(
        themes.map(file => fs.writeFile(
          join(dirname(file), `${basename(file, '.mjs')}.d.mts`),
          'import { ThemeRegistrationRaw } from \'@shikijs/core\';declare const reg: ThemeRegistrationRaw;export default reg',
          'utf-8',
        ),
        ),
      )
    },
  },
})
