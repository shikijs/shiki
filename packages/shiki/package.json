{
  "name": "shiki",
  "type": "module",
  "version": "1.14.1",
  "description": "A beautiful Syntax Highlighter.",
  "author": "Pine Wu <octref@gmail.com>; Anthony Fu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/shikijs/shiki#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/shikijs/shiki.git",
    "directory": "packages/shiki"
  },
  "bugs": "https://github.com/shikijs/shiki/issues",
  "keywords": [
    "shiki"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.mts",
      "default": "./dist/index.mjs"
    },
    "./core": {
      "types": "./dist/core.d.mts",
      "unwasm": "./dist/core-unwasm.mjs",
      "default": "./dist/core.mjs"
    },
    "./wasm": {
      "unwasm": "./dist/onig.wasm",
      "types": "./dist/wasm.d.mts",
      "default": "./dist/wasm.mjs"
    },
    "./langs": {
      "types": "./dist/langs.d.mts",
      "default": "./dist/langs.mjs"
    },
    "./themes": {
      "types": "./dist/themes.d.mts",
      "default": "./dist/themes.mjs"
    },
    "./theme-css-variables": {
      "types": "./dist/theme-css-variables.d.mts",
      "default": "./dist/theme-css-variables.mjs"
    },
    "./bundle/full": {
      "types": "./dist/bundle-full.d.mts",
      "default": "./dist/bundle-full.mjs"
    },
    "./bundle/web": {
      "types": "./dist/bundle-web.d.mts",
      "default": "./dist/bundle-web.mjs"
    },
    "./onig.wasm": "./dist/onig.wasm",
    "./dist/*": "./dist/*",
    "./package.json": "./package.json",
    "./*": "./dist/*"
  },
  "main": "./dist/index.mjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.mts",
  "typesVersions": {
    "*": {
      "core": [
        "./dist/core.d.mts"
      ],
      "wasm": [
        "./dist/wasm.d.mts"
      ],
      "langs": [
        "./dist/langs.d.mts"
      ],
      "themes": [
        "./dist/themes.d.mts"
      ],
      "bundle/full": [
        "./dist/bundle-full.d.mts"
      ],
      "bundle/web": [
        "./dist/bundle-web.d.mts"
      ],
      "theme-css-variables": [
        "./dist/theme-css-variables.d.mts"
      ],
      "*": [
        "./dist/*",
        "./*"
      ]
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "prepare": "esno scripts/prepare.ts",
    "build": "rimraf dist && NODE_OPTIONS=--max-old-space-size=4096 rollup -c",
    "dev": "rollup -cw",
    "prepublishOnly": "nr build",
    "test": "vitest",
    "test:cf": "wrangler dev test/cf.ts --port 60001"
  },
  "dependencies": {
    "@shikijs/core": "workspace:*",
    "@types/hast": "catalog:"
  },
  "devDependencies": {
    "tm-grammars": "catalog:",
    "tm-themes": "catalog:",
    "vscode-oniguruma": "catalog:"
  }
}
