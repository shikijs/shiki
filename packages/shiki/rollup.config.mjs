// @ts-check
import { basename, dirname, join } from 'node:path'
import commonjs from '@rollup/plugin-commonjs'
import json from '@rollup/plugin-json'
import { nodeResolve } from '@rollup/plugin-node-resolve'
import fg from 'fast-glob'
import fs from 'fs-extra'
import { defineConfig } from 'rollup'
import copy from 'rollup-plugin-copy'
import dts from 'rollup-plugin-dts'
import esbuild from 'rollup-plugin-esbuild'

const entries = [
  'src/index.ts',
  'src/core.ts',
  'src/core-unwasm.ts',
  'src/types.ts',
  'src/themes.ts',
  'src/langs.ts',
  'src/wasm.ts',
  'src/bundle-full.ts',
  'src/bundle-web.ts',
  'src/theme-css-variables.ts',
  'src/engine-javascript.ts',
  'src/engine-oniguruma.ts',
  'src/textmate.ts',
]

const external = [
  '@shikijs/types',
  '@shikijs/core',
  '@shikijs/core/wasm-inlined',
  '@shikijs/core/types',
  '@shikijs/engine-javascript',
  '@shikijs/engine-oniguruma',
  '@shikijs/vscode-textmate',
  'shiki/wasm',
  /[/\\]src[/\\](langs|themes)[/\\]/g,
]

const plugins = [
  esbuild(),
  nodeResolve(),
  commonjs(),
  json({
    namedExports: false,
    preferConst: true,
    compact: true,
  }),
]

export default defineConfig([
  {
    input: [
      ...entries,
    ],
    output: {
      dir: 'dist',
      format: 'esm',
      entryFileNames: '[name].mjs',
    },
    plugins: [
      ...plugins,
      copy({
        targets: [
          { src: './src/langs', dest: 'dist' },
          { src: './src/themes', dest: 'dist' },
        ],
      }),
    ],
    external,
  },
  {
    input: entries,
    output: {
      dir: 'dist',
      format: 'esm',
      chunkFileNames: 'types/[name].d.mts',
      entryFileNames: f => `${f.name.replace(/src[\\/]/, '')}.d.mts`,
    },
    plugins: [
      dts({
        respectExternal: true,
      }),
      copy({
        targets: [{ src: './node_modules/vscode-oniguruma/release/onig.wasm', dest: 'dist' }],
      }),
      {
        name: 'post',
        async buildEnd() {
          await fs.writeFile('dist/onig.d.mts', 'declare const binary: ArrayBuffer; export default binary;', 'utf-8')
          const langs = await fg('dist/langs/*.mjs', { absolute: true })
          await Promise.all(
            langs.map(file => fs.writeFile(
              join(dirname(file), `${basename(file, '.mjs')}.d.mts`),
              'import { LanguageRegistration } from \'@shikijs/core\';declare const reg: LanguageRegistration[];export default reg',
              'utf-8',
            )),
          )
          const themes = await fg('dist/themes/*.mjs', { absolute: true })
          await Promise.all(
            themes.map(file => fs.writeFile(
              join(dirname(file), `${basename(file, '.mjs')}.d.mts`),
              'import { ThemeRegistrationRaw } from \'@shikijs/core\';declare const reg: ThemeRegistrationRaw;export default reg',
              'utf-8',
            )),
          )
        },
      },
    ],
    external,
    onwarn: (warning, warn) => {
      if (!/Circular|an empty chunk/.test(warning.message))
        warn(warning)
    },
  },
])
