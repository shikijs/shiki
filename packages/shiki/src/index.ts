export * from './bundle-full'

export type { BuiltinLanguage, BuiltinTheme } from './types'
export { createJavaScriptRegexEngine, defaultJavaScriptRegexConstructor } from '@shikijs/engine-javascript'
export { createOnigurumaEngine, loadWasm } from '@shikijs/engine-oniguruma'
