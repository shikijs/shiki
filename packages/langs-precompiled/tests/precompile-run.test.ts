import { createJavaScriptRawEngine } from '@shikijs/engine-javascript'
import { createHighlighterCore } from 'shiki'
import { expect, it } from 'vitest'
// eslint-disable-next-line antfu/no-import-dist
import vitesseDark from '../../themes/dist/vitesse-dark.mjs'

const isNode20andUp = process.version.replace(/^v/, '').split('.').map(Number)[0] >= 20

it.runIf(isNode20andUp)('run', async () => {
  const shiki = await createHighlighterCore({
    themes: [vitesseDark],
    langs: [import('../dist/html.mjs')],
    engine: createJavaScriptRawEngine(),
  })

  const code = shiki.codeToHtml('<div>hello</div>', { lang: 'html', theme: 'vitesse-dark' })
  expect(code).toMatchSnapshot()
})
