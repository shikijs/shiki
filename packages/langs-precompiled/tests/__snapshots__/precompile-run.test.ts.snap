// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`run 1`] = `"<pre class="shiki vitesse-dark" style="background-color:#121212;color:#dbd7caee" tabindex="0"><code><span class="line"><span style="color:#666666">&#x3C;</span><span style="color:#4D9375">div</span><span style="color:#666666">></span><span style="color:#DBD7CAEE">hello</span><span style="color:#666666">&#x3C;/</span><span style="color:#4D9375">div</span><span style="color:#666666">></span></span></code></pre>"`;
